types/src/v0/impls/stake_table.rs

```diff
@@ -974,6 +981,136 @@ impl StakeTableFetcher {
+    /// - However, once we support reward withdrawals, new mints will likely also come from the zero address**,
+    ///   making this logic unsafe.
+    ///
+    /// Safer approach:
+    /// - We now ely on the `Initialized` event of the token contract (which should only occur once).
+    /// - After locating this event, we **fetch its transaction receipt and look for a decoded `Transfer` log** (mint).
+    /// - If either step fails, the function aborts to prevent incorrect reward calculations.
+    pub async fn fetch_block_reward(&self) -> anyhow::Result<RewardAmount> {
+        let max_events_range = self.l1_client.options().l1_events_max_block_range;
+        let chain_config = *self.chain_config.lock().await;
+
+        let Some(stake_table_contract) = chain_config.stake_table_contract else {
+            bail!("No stake table contract address found in Chain config");
+        };
+
+        let provider = self.l1_client.provider.clone();
+        let stake_table = StakeTableV2::new(stake_table_contract, provider.clone());
+
+        let stake_table_init_block = stake_table
+            .initializedAtBlock()
+            .block(BlockId::finalized())
+            .call()
+            .await?
+            ._0
+            .to::<u64>();
+
+        tracing::info!("stake table init block ={stake_table_init_block}");
+
+        let token_address = stake_table
+            .token()
+            .block(BlockId::finalized())
+            .call()
+            .await
+            .context("Failed to get token address")?
+            ._0;
+
+        let token = EspToken::new(token_address, provider.clone());
+
+        // Attempt to find Initialized event
+        let init_logs = token
+            .Initialized_filter()
+            .from_block(0u64)
+            .to_block(BlockNumberOrTag::Finalized)
+            .query()
+            .await?;
+
+        let initial_supply = if let Some((_, init_log)) = init_logs.first().cloned() {
+            // Get the transaction that emitted the Initialized event
+            let init_tx = provider
+                .get_transaction_receipt(init_log.transaction_hash.unwrap())
+                .await?
+                .context("Failed to get transaction for Initialized event")?;
+
+            let mint_transfer = init_tx
+                .decoded_log::<EspToken::Transfer>()
+                .context("Failed to decode Transfer log")?;
+
+            tracing::debug!("mint transfer event ={mint_transfer:?}");
+            ensure!(
+                mint_transfer.from == Address::ZERO,
+                "First transfer should be a mint from the zero address"
+            );
+
+            mint_transfer.value
+        } else {
+            const MAX_BLOCKS_SCANNED: u64 = 200_000;
+            let mut total_scanned = 0;
+
+            tracing::error!("No Initialized event found for fetching block reward");
+
+            // Fallback: manually search for mint transfer (from zero address)
```

Comment:
I think we should scan for the Initialized event, not the mint transfer. Then get the mint from the txn receipt.